{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Worksheet 02 - Annotating Genomes With `Prokka`, and Calculating a Pangenome with `Roary` <img src=\"images/JHI_STRAP_Web.png\" style=\"width: 150px; float: right;\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "In this worksheet, you will use [`Prokka`]() and [`Roary`](https://github.com/sanger-pathogens/Roary) to annotate and calculate a pangenome for assemblies of the bacterial plant pathogen *Pantoea agglomerans*. \n",
    "\n",
    "The `Prokka` software conducts rapid annotation of microbial genomes, and the `Roary` software enables rapid pangenome identification in a tool that can be run on a desktop or laptop computer."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running cells in this notebook\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\">\n",
    "This is an interactive notebook, which means you are able to run the code that is written in each of the cells.\n",
    "<br /><br />\n",
    "To run the code in a cell, you should:\n",
    "<br /><br />\n",
    "<ol>\n",
    "<li>Place your mouse cursor in the cell, and click (this gives the cell *focus*) to make it active \n",
    "<li>Hold down the `Shift` key, and press the `Return` key.\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "If this is successful, you should see the input marker to the left of the cell change from\n",
    "\n",
    "```\n",
    "In [ ]:\n",
    "```\n",
    "\n",
    "to (for example)\n",
    "\n",
    "```\n",
    "In [1]:\n",
    "```\n",
    "\n",
    "and you may see output appear below the cell."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Related online documentation\n",
    "\n",
    "**Publications**\n",
    "\n",
    "* \"Roary: Rapid large-scale prokaryote pan genome analysis\", Andrew J. Page, Carla A. Cummins, Martin Hunt, Vanessa K. Wong, Sandra Reuter, Matthew T. G. Holden, Maria Fookes, Daniel Falush, Jacqueline A. Keane, Julian Parkhill,\n",
    "*Bioinformatics*, (2015). [doi:10.1093/bioinformatics/btv421](http://dx.doi.org/10.1093/bioinformatics/btv421)\n",
    "\n",
    "\n",
    "**Software**\n",
    "\n",
    "* `pyani` homepage: [http://widdowquinn.github.io/pyani/](http://widdowquinn.github.io/pyani/)\n",
    "* `Roary` software: [https://github.com/sanger-pathogens/Roary](https://github.com/sanger-pathogens/Roary)\n",
    "* `Prokka` homepage: [http://www.vicbioinformatics.com/software.prokka.shtml](http://www.vicbioinformatics.com/software.prokka.shtml)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Requirements\n",
    "\n",
    "<div class=\"alert alert-success\">\n",
    "To complete this worksheet, you will need:\n",
    "<ul>\n",
    "<li>an active internet connection\n",
    "<li>a local, working installation of <b>pyani</b>\n",
    "<li>a local, working installation of <b>Prokka</b>\n",
    "<li>a local, working installation of <b>Roary</b>\n",
    "</ul>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Downloading *Pantoea agglomerans* assemblies\n",
    "\n",
    "We will use the `genbank_get_genomes_by_taxon.py` script, part of `pyani` to download all available *Pantoea agglomerans* assembly sequences to the subdirectory `Pantoea_agglomerans`. To do this, issue the following command at the terminal:\n",
    "\n",
    "```\n",
    "genbank_get_genomes_by_taxon.py -v --format fasta --email <your_email> \\\n",
    "  -o Pantoea_agglomerans -t 549\n",
    "```\n",
    "\n",
    "This should generate verbose output describing the operation of downloading (at least) 15 genome assembly sequences. Once the command is complete, you should see the following (or similar) contents of the `Pantoea_agglomerans` subdirectory when issuing the `tree` command:\n",
    "\n",
    "```\n",
    "$ tree Pantoea_agglomerans/\n",
    "Pantoea_agglomerans/\n",
    "├── GCF_000241285.1.fasta\n",
    "├── GCF_000330765.1.fasta\n",
    "├── GCF_000475055.1.fasta\n",
    "├── GCF_000627115.1.fasta\n",
    "├── GCF_000687245.1.fasta\n",
    "├── GCF_000710215.1.fasta\n",
    "├── GCF_000731125.1.fasta\n",
    "├── GCF_000743785.1.fasta\n",
    "├── GCF_000743785.2.fasta\n",
    "├── GCF_000757415.1.fasta\n",
    "├── GCF_000814075.1.fasta\n",
    "├── GCF_001288285.1.fasta\n",
    "├── GCF_001558735.1.fasta\n",
    "├── GCF_001597625.1.fasta\n",
    "├── GCF_001598475.1.fasta\n",
    "├── classes.txt\n",
    "└── labels.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Fixing downloaded files for use with `Prokka`\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "The assembly FASTA files downloaded from NCBI may not work directly with <b>Prokka</b>, as their contig names are too long (there is a 20 character limit), and give the following error:\n",
    "</div>\n",
    "\n",
    "```\n",
    "Contig ID must <= 20 chars long: gi|358545120|dbj|BAEF01000003.1|\n",
    "```\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "To avoid this, we will rename all our downloaded files *in place* (i.e. without copying and rewriting) with unique ID numbers, from 1 to $n$, where $n$ is the number of contigs in the FASTA file. This is done by executing the cell below.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# Use Biopython for sequence reading/writing\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Input directory, containing FASTA files\n",
    "indir = \"Pantoea_agglomerans\"\n",
    "\n",
    "# Rename all contigs in each FASTA file of the input directory, \n",
    "# from 1 to n, where n is the number of contigs in the FASTA file\n",
    "# Loop over all FASTA files in the directory, and rename contigs\n",
    "# in-place, in each file.\n",
    "for filename in [fname for fname in os.listdir(indir)\n",
    "                 if os.path.splitext(fname)[-1] == '.fasta']:\n",
    "    infile = os.path.join(indir, filename)\n",
    "    seqrecords = list(SeqIO.parse(infile, 'fasta'))\n",
    "    for idx, record in enumerate(seqrecords):\n",
    "        record.id = str(idx)\n",
    "    SeqIO.write(seqrecords, infile, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When the cell above has finished running, we have an input dataset that is suitable for running `Prokka`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Annotating genomes with `Prokka`\n",
    "\n",
    "We will use `Prokka` with default settings to annotate the downloaded assemblies.\n",
    "\n",
    "For a single input sequences, such as `Pantoea_agglomerans/GCF_000241285.1.fasta`, a simple default command is:\n",
    "\n",
    "```\n",
    "prokka Pantoea_agglomerans/GCF_000241285.1.fasta --genus Pantoea --species agglomerans\n",
    "```\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\">\n",
    "On my laptop, this takes ≈10min to annotate the genome above so, to annotate 15 genomes, we could expect this exercise to take approximately 2.5 hours. Because of this, the <b>Prokka</b> <b>.gff</b> output for 15 *P. agglomerans* genomes has been prepared using the script <b>scripts/02-prokka.sh</b>, and the output placed in the subdirectory <b>prokka_out</b>.\n",
    "</div>\n",
    "\n",
    "This script can be run by issuing the following command at the terminal:\n",
    "\n",
    "```\n",
    "sh scripts/02-prokka.sh\n",
    "```\n",
    "\n",
    "It will generate one subdirectory under `prokka_out` for each input sequence:\n",
    "\n",
    "```\n",
    "$ tree -L 1 prokka_out/\n",
    "prokka_out/\n",
    "├── GCF_000241285.1\n",
    "├── GCF_000330765.1\n",
    "├── GCF_000475055.1\n",
    "├── GCF_000627115.1\n",
    "├── GCF_000687245.1\n",
    "├── GCF_000710215.1\n",
    "├── GCF_000731125.1\n",
    "├── GCF_000743785.1\n",
    "├── GCF_000743785.2\n",
    "├── GCF_000757415.1\n",
    "├── GCF_000814075.1\n",
    "├── GCF_001288285.1\n",
    "├── GCF_001558735.1\n",
    "├── GCF_001597625.1\n",
    "└── GCF_001598475.1\n",
    "```\n",
    "\n",
    "and each subdirectory will contain the annotation for that input sequence:\n",
    "\n",
    "```\n",
    "$ tree prokka_out/GCF_000241285.1/\n",
    "prokka_out/GCF_000241285.1/\n",
    "├── PROKKA_05212016.err\n",
    "├── PROKKA_05212016.faa\n",
    "├── PROKKA_05212016.ffn\n",
    "├── PROKKA_05212016.fna\n",
    "├── PROKKA_05212016.fsa\n",
    "├── PROKKA_05212016.gbk\n",
    "├── PROKKA_05212016.gff\n",
    "├── PROKKA_05212016.log\n",
    "├── PROKKA_05212016.sqn\n",
    "├── PROKKA_05212016.tbl\n",
    "└── PROKKA_05212016.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Constructing a pangenome with `Roary`\n",
    "\n",
    "We will construct a default pangenome for *P. agglomerans* using `Roary`, by issuing the following command at the terminal: \n",
    "\n",
    "```\n",
    "roary -f roary_out -v prokka_out/*/*.gff\n",
    "```\n",
    "\n",
    "The `-f` flag places `Roary`'s output in the subdirectory `roary_<something>`, where `<something>` can vary from run to run:\n",
    "\n",
    "```\n",
    "$ tree roary/\n",
    "roary/\n",
    "├── accessory.header.embl\n",
    "├── accessory.tab\n",
    "├── accessory_binary_genes.fa\n",
    "├── accessory_binary_genes.fa.newick\n",
    "├── accessory_graph.dot\n",
    "├── blast_identity_frequency.Rtab\n",
    "├── clustered_proteins\n",
    "├── core_accessory.header.embl\n",
    "├── core_accessory.tab\n",
    "├── core_accessory_graph.dot\n",
    "├── fixed_input_files\n",
    "│   ├── GCF_000330765.1.gff\n",
    "│   ├── GCF_000475055.1.gff\n",
    "│   ├── GCF_000627115.1.gff\n",
    "│   ├── GCF_000687245.1.gff\n",
    "│   └── GCF_000710215.1.gff\n",
    "├── gene_presence_absence.Rtab\n",
    "├── gene_presence_absence.csv\n",
    "├── number_of_conserved_genes.Rtab\n",
    "├── number_of_genes_in_pan_genome.Rtab\n",
    "├── number_of_new_genes.Rtab\n",
    "├── number_of_unique_genes.Rtab\n",
    "└── summary_statistics.txt\n",
    "```\n",
    "\n",
    "`Roary` defines (arbitrary) cutoffs for four classes of gene, representing different levels of conservation across the input genome set:\n",
    "\n",
    "* Core genes (99-100% of input strains)\n",
    "* Soft core genes (95-99%)\n",
    "* Shell genes\t(15-95%)\n",
    "* Cloud genes\t(0-15%)\n",
    "\n",
    "The numbers of genes placed in each class are summarised in the `summary_statistics.txt` file, and several useful output files are also produced."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  },
  "widgets": {
   "state": {},
   "version": "1.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
